python/ray/data/_internal/logical/rules/inherit_batch_format.py

```diff
@@ -0,0 +1,42 @@
+from collections import deque
+from typing import Iterable
+
+from ray.data._internal.logical.interfaces import LogicalOperator, LogicalPlan, Rule
+from ray.data._internal.logical.operators.all_to_all_operator import Aggregate, Sort
+from ray.data._internal.logical.operators.map_operator import MapBatches
+
+
+class InheritBatchFormatRule(Rule):
+    """For Sort and Aggregate logical operator, apply this rule
+    to inherit batch_format from upstream operator by traversing
+    the entire DAG."""
+
+    def apply(self, plan: LogicalPlan) -> LogicalPlan:
+        optimized_dag: LogicalOperator = self._apply(plan.dag)
+        new_plan = LogicalPlan(dag=optimized_dag, context=plan.context)
+        return new_plan
+
+    def _apply(self, op: LogicalOperator):
+        # Post-order traversal.
+        nodes: Iterable[LogicalOperator] = deque()
+        for node in op.post_order_iter():
+            nodes.appendleft(node)
+
+        while len(nodes) > 0:
+            current_op = nodes.pop()
+
+            if isinstance(current_op, (Sort, Aggregate)):
```

Comment:
@xingyu-long let's generalize this rule to be propagate this across the board to all ops not just Sort/Aggregate

Reply:
@alexeykudinkin thanks for the suggestion! are we thinking have `._batch_format` for all ops implicitly? i.e., even for some op, we don't have `._batch_format` field? @scottjlee  could you give some thoughts on this?

Reply:
i think it would make sense to extend this to all sub-operators of `AllToAllOperator`, and add `._batch_format` accordingly. (your PR already does this) the only operator we can inherit the batch format is `MapBatches` operator (for example, `MapRows` wouldn't have any notion of a "batch format"), or other upstream `AllToAllOperator`s. so we can update the condition here to check for any `AllToAllOperator` and inherit the batch format, instead of just `Sort` and `Aggregate`.

Reply:
I see. makes sense. but btw, I think it should be `AbstractAlltoAll` since we are using logical operator here. @scottjlee @alexeykudinkin  could you take a final look at the changes? thanks!